network/server.go

```diff
@@ -237,70 +170,153 @@ func NewServer(logger hclog.Logger, config *Config) (*Server, error) {
+		}
 
-		// try to decode the bootnodes
-		bootnodes := []*peer.AddrInfo{}
+		key = libp2pKey
+	}
 
-		for _, raw := range s.config.Chain.Bootnodes {
-			node, err := StringToAddrInfo(raw)
-			if err != nil {
-				return fmt.Errorf("failed to parse bootnode %s: %w", raw, err)
-			}
+	return key, nil
+}
 
-			if node.ID == s.host.ID() {
-				s.logger.Info("Omitting bootnode with same ID as host", "id", node.ID)
+// Start starts the networking services
+func (s *Server) Start() error {
+	s.logger.Info("LibP2P server running", "addr", common.AddrInfoToString(s.AddrInfo()))
 
-				continue
-			}
+	if setupErr := s.setupIdentity(); setupErr != nil {
+		return fmt.Errorf("unable to setup identity, %w", setupErr)
+	}
 
-			bootnodes = append(bootnodes, node)
+	// Set up the peer discovery mechanism if needed
+	if !s.config.NoDiscover {
+		// Parse the bootnode data
+		if setupErr := s.setupBootnodes(); setupErr != nil {
+			return fmt.Errorf("unable to parse bootnode data, %w", setupErr)
 		}
 
-		if setupErr := s.discovery.setup(bootnodes); setupErr != nil {
+		// Setup and start the discovery service
+		if setupErr := s.setupDiscovery(); setupErr != nil {
 			return fmt.Errorf("unable to setup discovery, %w", setupErr)
 		}
 	}
 
-	go func() {
-		if err := s.runJoinWatcher(); err != nil {
-			s.logger.Error(fmt.Sprintf("Unable to start join watcher service, %v", err))
-		}
-	}()
+	go s.runDial()
+	go s.checkPeerConnections()
 
 	// watch for disconnected peers
 	s.host.Network().Notify(&network.NotifyBundle{
 		DisconnectedF: func(net network.Network, conn network.Conn) {
-			go func() {
-				s.delPeer(conn.RemotePeer())
-			}()
+			// Update the local connection metrics
+			s.removePeer(conn.RemotePeer())
 		},
 	})
 
 	return nil
 }
 
+// setupBootnodes sets up the node's bootnode connections
+func (s *Server) setupBootnodes() error {
+	// Check the bootnode config is present
+	if s.config.Chain.Bootnodes == nil {
+		return ErrNoBootnodes
+	}
+
+	// Check if at least one bootnode is specified
+	if len(s.config.Chain.Bootnodes) < MinimumBootNodes {
+		return ErrMinBootnodes
+	}
+
+	bootnodesArr := make([]*peer.AddrInfo, 0)
+	bootnodesMap := make(map[peer.ID]*peer.AddrInfo)
+
+	for _, rawAddr := range s.config.Chain.Bootnodes {
+		bootnode, err := common.StringToAddrInfo(rawAddr)
+		if err != nil {
+			return fmt.Errorf("failed to parse bootnode %s: %w", rawAddr, err)
+		}
+
+		if bootnode.ID == s.host.ID() {
+			s.logger.Info("Omitting bootnode with same ID as host", "id", bootnode.ID)
+
+			continue
+		}
+
+		bootnodesArr = append(bootnodesArr, bootnode)
+		bootnodesMap[bootnode.ID] = bootnode
+	}
+
+	// It's fine for the bootnodes field to be unprotected
+	// at this point because it is initialized once (doesn't change),
+	// and used only after this point
+	s.bootnodes = &bootnodesWrapper{
+		bootnodeArr:       bootnodesArr,
+		bootnodesMap:      bootnodesMap,
+		bootnodeConnCount: int64(len(bootnodesArr)),
```

Comment:
Why bootnodeConnCount has value as default? This is not the number of connecting bootnode?

Reply:
It's 0 by default when the server is **created**, but at this point where the value is being set, it is the number of successful bootnode connections. Please check the point of calling `setupBootnodes`, and what it does, it will be much clearer 🙏

Reply:
@zivkovicmilos So why you increment/decrement this count by `increaseBootnodeConnCount` when connecting/disconnecting?

Reply:
The initial value for this counter is `0`, when the server is created.
So there are essentially 2 places where we update this counter:
1. When trying to establish the _initial_ bootnode connections (method `setupBootnodes`). We update the counter with the number of _active_ (successful) bootnode connections
2. When a peer connects / disconnects, and they were part of the bootnode set. We update the number of active bootnode connections then ([here](https://github.com/0xPolygon/polygon-edge/blob/449cd90d57247c3b5f8501094fba2dc632af0e2a/network/server_identity.go#L77) and [here](https://github.com/0xPolygon/polygon-edge/blob/449cd90d57247c3b5f8501094fba2dc632af0e2a/network/server.go#L501))

Reply:
I just noticed the problem you mentioned 🤦 

Bootnode conn counts should be updated on connections **only**, and not on this initialization (parsing) 🤦 

Fixed it here 👍 :

[c6de5d7](https://github.com/0xPolygon/polygon-edge/pull/343/commits/c6de5d7e89bcf44ac2f52ffb4a912c82f9640052)